pkg/abstractions/common/ring_buffer.go

```diff
@@ -20,19 +20,34 @@ func NewRingBuffer[T any](size int) *RingBuffer[T] {
 	}
 }
 
-// Push adds a new request to the buffer
-func (rb *RingBuffer[T]) Push(request T) {
+// Push adds a new request to the buffer. If priority is true, it gets inserted at the front of the buffer
+func (rb *RingBuffer[T]) Push(request T, priority bool) {
 	rb.mu.Lock()
 	defer rb.mu.Unlock()
 
-	if rb.count == rb.size {
-		rb.head = (rb.head + 1) % rb.size
+	if priority {
+		if rb.count == rb.size {
+			// Buffer is full, overwrite the oldest element at head without moving pointers
+			// No adjustment to head or tail, since the count remains the same
+			rb.buffer[rb.head] = request
+		} else {
+			// Move head backward to insert at front
+			rb.head = (rb.head - 1 + rb.size) % rb.size
+			rb.buffer[rb.head] = request
+			rb.count++
+		}
```

Comment:
If priority is for failed requests and tons of requests start failing it seems like we'll just start dropping a lot of those. Is that ok? Not sure if I'm misunderstanding the use case/if that's intended.

Reply:
The current behavior is the failed request will be inserted at the back of the FIFO. The failure mode we're improving here is that a container was temporarily busy, and we want to rebuffer that request with high priority so its handled as soon as possible (since it was requested first). But yes, if tons of requests are failing, tons of requests will still fail. Not much we can do about that.